### PR TITLE
Cancel actions

### PR DESCRIPTION
**Pull request by oscar-lima**
## Brief

This PR updates Grasplan to listen to preemption requests from the user and act accordingly. Specifically, preemption requests are now forwarded to MoveIt.

NOTE: Not (yet) tested on real robot.

## Issue Encountered:

While MoveIt implements a preemption callback in its pick_place_action_capability.cpp (reference: https://github.com/moveit/moveit/blob/noetic-devel/moveit_ros/manipulation/move_group_pick_place_capability/src/pick_place_action_capability.cpp#L405), the callback is currently empty, resulting in MoveIt ignoring preemption requests. This behavior prevents MoveIt from halting ongoing actions when preempted by the user.

## Workaround Implemented:

To address this limitation, we introduced a custom canceller that directly hijacks the arm and gripper controllers. The custom canceller achieves the following:

- Tracks Incoming Goals: It monitors all controller goals sent by MoveIt to the arm and gripper controllers.
- Sends Preemption Requests: For each tracked goal, the canceller issues a preemption request to the controllers. This effectively halts the arm and prevents further movement.

## Conclusion

This approach ensures that preemption requests are honored, even when MoveIt itself fails to handle them properly.

### CODE REVIEW

File: src/grasplan/place.py [hunk 1]
```diff
@@ -278,6 +287,7 @@ def place_object(
         '''
         create action lib client and call moveit place action server
         '''
+        PLACE_OBJECT_SERVER_NAME = 'place'
```
Comment (oscar-lima) [inline src/grasplan/place.py:308]: I am sneaking in some refactoring to make it even with pick node

File: src/grasplan/place.py [hunk 2]
```diff
@@ -358,12 +367,17 @@ def place_object(
...
 
-            rospy.loginfo(f'sending place {object_to_be_placed} goal to {self.place_object_server_name} action server')
-            action_client.send_goal(goal)
-            rospy.loginfo(f'waiting for result from {self.place_object_server_name} action server')
-            if action_client.wait_for_result(rospy.Duration.from_sec(self.timeout)):
+            rospy.loginfo(
+                f'sending place {object_to_be_placed} goal to {PLACE_OBJECT_SERVER_NAME} action server '
+                'and waiting for result'
+            )
+            if self.action_client_helper.send_goal_to_rogue_server_and_wait(goal, action_client, patience_timeout=0.1):
```
Comment (oscar-lima) [inline src/grasplan/place.py]: This is a combination of "action_client.send_goal(goal)" and "action_client.wait_for_result" that checks at 10 hz if preemption was requested by user, this is done in a new file called action_client_helper.py . Additionally if MoveIt ignores the cancellation request, it forces it. patience_timeout is the amount of time to wait for MoveIt to acknowledge the cancellation. 0.1 is quite low but we are factoring in expert knowledge in the sense that we already know that our request will be ignored... and therefore it does not make sense to make this value higher.

File: src/grasplan/tools/action_client_helper.py [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        Example:
+            If `ns` is '/robot_namespace' and `controller_names` is ['arm', 'gripper'],
+            the trajectory canceller topics will be:
+                - /robot_namespace/arm_controller/follow_joint_trajectory/cancel
+                - /robot_namespace/gripper_controller/follow_joint_trajectory/cancel
+        """
+        self.canceller_list = []
+        for controller_name in controller_names:
+            self.canceller_list.append(
+                TrajectoryCanceller(f'/{ns}/{controller_name}_controller/follow_joint_trajectory/')
```
Comment (oscar-lima) [inline src/grasplan/tools/action_client_helper.py:134]: NOTE: two object instances are created, one for the arm and one for the gripper, however note that this is done 1 time for pick, 1 time for place and 1 for insert, making in total 6 subscribers. This is not ideal... but unfortunately an optimal solution of having 2 cancellers (one for arm and one for gripper) that is shared between the pick, place and insert modules, would require, option 1) heavy refactoring of Grasplan code or 2) having the canceller as a separate node and making service calls to it. For now I leave it like this, but we can have a more in-depth discussion about it.

**Comment by oscar-lima** [comment]
Comment from Marc: When canceling a placement operation in a cluttered table setup and place module is forced to go into retry stages 2 or/and 3, the force canceling process starts but fails after 100 attempts, causing MoveIt to continue placing.

**Comment by mintar** [comment]
Alternatives:

* implement the preemption callback inside MoveIt so that it properly handles cancel requests, send them a pull request 
* switch from the old pick and place pipeline to the MoveIt Task Constructor

**Comment by oscar-lima** [comment]
> Comment from Marc: When canceling a placement operation in a cluttered table setup and place module is forced to go into retry stages 2 or/and 3, the force canceling process starts but fails after 100 attempts, causing MoveIt to continue placing.

place goals were splitted into batches of configurable size to resolve this issue, see: 1abee2b9875e6b809eb62957dabfd3c184cb2211

**Comment by oscar-lima** [comment]
> Alternatives:
> 
>     * implement the preemption callback inside MoveIt so that it properly handles cancel requests, send them a pull request
> 
>     * switch from the old pick and place pipeline to the MoveIt Task Constructor

Thank you for the feedback! Since this is part of older code that's been replaced by MTC, I’m hesitant to submit a PR at this stage. Migrating to MTC is definitely the right direction, but I plan to hold off for now as I’m currently gathering requirements for a major Grasplan 2.0 refactoring.

**Comment by oscar-lima** [comment]
I was holding on this PR as a draft because the action test clients for place and insert were not updated, which is now resolved by 2588ce783b4340fa1e23d15cffb44bf4971f3827

**Pull request merged**